File: backend/app/models/backtest.py
"""Backtesting models."""

from sqlalchemy import Column, String, Float, Integer, DateTime, ForeignKey, Enum, Boolean, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy import JSON
from sqlalchemy.orm import relationship
import uuid
from datetime import datetime
import enum

from app.db.database import Base


class BacktestStatus(str, enum.Enum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class Backtest(Base):
    __tablename__ = "backtests"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    strategy_id = Column(UUID(as_uuid=True), ForeignKey("trading_strategies.id"), nullable=False)
    
    # Backtest configuration
    name = Column(String, nullable=False)
    description = Column(Text)
    start_date = Column(DateTime, nullable=False)
    end_date = Column(DateTime, nullable=False)
    initial_capital = Column(Float, nullable=False)
    
    # Transaction costs
    commission = Column(Float, default=0)
    slippage = Column(Float, default=0)
    
    # Status
    status = Column(Enum(BacktestStatus), default=BacktestStatus.PENDING)
    progress = Column(Float, default=0)  # 0-100%
    error_message = Column(Text)
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    started_at = Column(DateTime)
    completed_at = Column(DateTime)
    
    # Relationships
    user = relationship("User")
    strategy = relationship("app.models.trading.Strategy")
    results = relationship("BacktestResult", back_populates="backtest", uselist=False)
    trades = relationship("BacktestTrade", back_populates="backtest")


class BacktestResult(Base):
    __tablename__ = "backtest_results"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    backtest_id = Column(UUID(as_uuid=True), ForeignKey("backtests.id"), nullable=False, unique=True)
    strategy_id = Column(UUID(as_uuid=True), ForeignKey("trading_strategies.id"), nullable=False)
    
    # Performance metrics
    total_return = Column(Float)
    annualized_return = Column(Float)
    volatility = Column(Float)
    sharpe_ratio = Column(Float)
    sortino_ratio = Column(Float)
    calmar_ratio = Column(Float)
    
    # Risk metrics
    max_drawdown = Column(Float)
    max_drawdown_duration = Column(Integer)  # Days
    var_95 = Column(Float)  # Value at Risk 95%
    cvar_95 = Column(Float)  # Conditional VaR 95%
    
    # Trading metrics
    total_trades = Column(Integer)
    winning_trades = Column(Integer)
    losing_trades = Column(Integer)
    win_rate = Column(Float)
    profit_factor = Column(Float)
    expectancy = Column(Float)
    
    # Additional metrics
    beta = Column(Float)
    alpha = Column(Float)
    information_ratio = Column(Float)
    
    # Time series data (stored as JSON)
    equity_curve = Column(JSON)  # [{date, value}, ...]
    drawdown_curve = Column(JSON)
    returns_series = Column(JSON)
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    
    # Relationships
    backtest = relationship("Backtest", back_populates="results")
    strategy = relationship("app.models.trading.Strategy", back_populates="backtest_results")


class BacktestTrade(Base):
    __tablename__ = "backtest_trades"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    backtest_id = Column(UUID(as_uuid=True), ForeignKey("backtests.id"), nullable=False)
    
    # Trade details
    symbol = Column(String, nullable=False)
    side = Column(String, nullable=False)  # buy/sell
    quantity = Column(Float, nullable=False)
    entry_price = Column(Float, nullable=False)
    exit_price = Column(Float)
    
    # Dates
    entry_date = Column(DateTime, nullable=False)
    exit_date = Column(DateTime)
    
    # P&L
    gross_pnl = Column(Float)
    commission = Column(Float)
    slippage = Column(Float)
    net_pnl = Column(Float)
    
    # Trade metadata
    entry_reason = Column(String)
    exit_reason = Column(String)
    holding_period = Column(Integer)  # Days
    
    # Relationships
    backtest = relationship("Backtest", back_populates="trades")


class OptimizationRun(Base):
    __tablename__ = "optimization_runs"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    strategy_id = Column(UUID(as_uuid=True), ForeignKey("trading_strategies.id"), nullable=False)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    
    # Optimization configuration
    name = Column(String, nullable=False)
    parameters_to_optimize = Column(JSON)  # {param_name: {min, max, step}}
    optimization_metric = Column(String, nullable=False)  # sharpe_ratio, total_return, etc.
    method = Column(String, nullable=False)  # grid_search, genetic, bayesian
    
    # Date range
    start_date = Column(DateTime, nullable=False)
    end_date = Column(DateTime, nullable=False)
    
    # Status
    status = Column(Enum(BacktestStatus), default=BacktestStatus.PENDING)
    progress = Column(Float, default=0)
    total_iterations = Column(Integer)
    completed_iterations = Column(Integer, default=0)
    
    # Results
    optimal_parameters = Column(JSON)
    optimal_metric_value = Column(Float)
    
    # Performance surface (for visualization)
    performance_surface = Column(JSON)
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    started_at = Column(DateTime)
    completed_at = Column(DateTime)
    
    # Relationships
    strategy = relationship("app.models.trading.Strategy")
    user = relationship("User")
    results = relationship("OptimizationResult", back_populates="optimization_run")


class OptimizationResult(Base):
    __tablename__ = "optimization_results"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    optimization_run_id = Column(UUID(as_uuid=True), ForeignKey("optimization_runs.id"), nullable=False)
    
    # Parameters tested
    parameters = Column(JSON, nullable=False)
    
    # Performance metrics
    metric_value = Column(Float, nullable=False)  # The optimization metric
    total_return = Column(Float)
    sharpe_ratio = Column(Float)
    max_drawdown = Column(Float)
    win_rate = Column(Float)
    
    # Iteration info
    iteration_number = Column(Integer)
    
    # Relationships
    optimization_run = relationship("OptimizationRun", back_populates="results")


class WalkForwardAnalysis(Base):
    __tablename__ = "walk_forward_analyses"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    strategy_id = Column(UUID(as_uuid=True), ForeignKey("trading_strategies.id"), nullable=False)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    
    # Configuration
    name = Column(String, nullable=False)
    start_date = Column(DateTime, nullable=False)
    end_date = Column(DateTime, nullable=False)
    in_sample_periods = Column(Integer, nullable=False)  # Days
    out_sample_periods = Column(Integer, nullable=False)  # Days
    optimization_metric = Column(String, nullable=False)
    
    # Status
    status = Column(Enum(BacktestStatus), default=BacktestStatus.PENDING)
    current_period = Column(Integer, default=0)
    total_periods = Column(Integer)
    
    # Aggregate results
    avg_in_sample_performance = Column(Float)
    avg_out_sample_performance = Column(Float)
    performance_degradation = Column(Float)  # Difference between IS and OOS
    
    # Period results (stored as JSON)
    period_results = Column(JSON)
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    started_at = Column(DateTime)
    completed_at = Column(DateTime)
    
    # Relationships
    strategy = relationship("app.models.trading.Strategy")
    user = relationship("User")

File: frontend/src/services/api/backtest.ts
export type BacktestStatus = 'PENDING' | 'RUNNING' | 'COMPLETED' | 'FAILED' | 'CANCELLED'
export type StrategyType = 'MA_CROSSOVER' | 'RSI' | 'BOLLINGER_BANDS' | 'MACD' | 'CUSTOM'

export interface BacktestConfig {
  id: string
  name: string
  description?: string
  strategy: string
  strategyType?: StrategyType
  symbols: string[]
  startDate: string
  endDate: string
  initialCapital: number
  positionSize: number
  positionSizing?: 'FIXED' | 'PERCENT' | 'KELLY'
  maxPositions: number
  commission: number
  slippage: number
  stopLoss?: number
  takeProfit?: number
  parameters: Record<string, any>
  userId: string
  createdAt: string
  updatedAt: string
}

export interface BacktestResult {
  id: string
  configId: string
  config?: BacktestConfig
  status: BacktestStatus
  startedAt: string
  completedAt?: string
  duration?: number
  error?: string
  // Performance metrics
  totalReturn: number
  totalReturnPercent: number
  annualizedReturn: number
  sharpeRatio: number
  sortinoRatio: number
  calmarRatio: number
  maxDrawdown: number
  maxDrawdownPercent: number
  maxDrawdownDuration: number
  // Trade statistics
  totalTrades: number
  winningTrades: number
  losingTrades: number
  winRate: number
  avgWin: number
  avgLoss: number
  profitFactor: number
  expectancy: number
  avgTradeDuration: number
  // Risk metrics
  volatility: number
  var95: number
  cvar95: number
  beta?: number
  alpha?: number
  // Additional data
  equity: number[]
  drawdown: number[]
  trades: BacktestTrade[]
  dailyReturns: number[]
  monthlyReturns: Record<string, number>
  yearlyReturns: Record<string, number>
}

export interface BacktestTrade {
  id: string
  symbol: string
  entryDate: string
  exitDate: string
  entryPrice: number
  exitPrice: number
  quantity: number
  side: 'LONG' | 'SHORT'
  pnl: number
  pnlPercent: number
  commission: number
  slippage: number
  duration: number
  exitReason: 'SIGNAL' | 'STOP_LOSS' | 'TAKE_PROFIT' | 'END_OF_DATA'
}

export interface BacktestProgress {
  id: string
  status: BacktestStatus
  progress: number
  currentDate?: string
  message?: string
  tradesProcessed?: number
  estimatedTimeRemaining?: number
}

export interface BacktestComparison {
  configs: BacktestConfig[]
  results: BacktestResult[]
  metrics: {
    metric: string
    values: number[]
    winner: number
  }[]
}

export interface StrategyTemplate {
  id: string
  name: string
  description: string
  type: StrategyType
  defaultParameters: Record<string, any>
  requiredData: string[]
}

class BacktestService {
  async getConfigs(): Promise<BacktestConfig[]> {
    const response = await fetch('/api/v1/backtest/configs')
    if (!response.ok) throw new Error('Failed to fetch backtest configs')
    return response.json()
  }

  async getConfig(id: string): Promise<BacktestConfig> {
    const response = await fetch(`/api/v1/backtest/configs/${id}`)
    if (!response.ok) throw new Error('Failed to fetch backtest config')
    return response.json()
  }

  async createConfig(config: Partial<BacktestConfig>): Promise<BacktestConfig> {
    const response = await fetch('/api/v1/backtest/configs', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(config)
    })
    if (!response.ok) throw new Error('Failed to create backtest config')
    return response.json()
  }

  async updateConfig(id: string, updates: Partial<BacktestConfig>): Promise<BacktestConfig> {
    const response = await fetch(`/api/v1/backtest/configs/${id}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(updates)
    })
    if (!response.ok) throw new Error('Failed to update backtest config')
    return response.json()
  }

  async deleteConfig(id: string): Promise<void> {
    const response = await fetch(`/api/v1/backtest/configs/${id}`, {
      method: 'DELETE'
    })
    if (!response.ok) throw new Error('Failed to delete backtest config')
  }

  async runBacktest(configId: string): Promise<string> {
    const response = await fetch(`/api/v1/backtest/run/${configId}`, {
      method: 'POST'
    })
    if (!response.ok) throw new Error('Failed to start backtest')
    const { backtestId } = await response.json()
    return backtestId
  }

  async stopBacktest(backtestId: string): Promise<void> {
    const response = await fetch(`/api/v1/backtest/stop/${backtestId}`, {
      method: 'POST'
    })
    if (!response.ok) throw new Error('Failed to stop backtest')
  }

  async getBacktestStatus(backtestId: string): Promise<BacktestProgress> {
    const response = await fetch(`/api/v1/backtest/status/${backtestId}`)
    if (!response.ok) throw new Error('Failed to fetch backtest status')
    return response.json()
  }

  async getResults(): Promise<BacktestResult[]> {
    const response = await fetch('/api/v1/backtest/results')
    if (!response.ok) throw new Error('Failed to fetch backtest results')
    return response.json()
  }

  async getResult(id: string): Promise<BacktestResult> {
    const response = await fetch(`/api/v1/backtest/results/${id}`)
    if (!response.ok) throw new Error('Failed to fetch backtest result')
    return response.json()
  }

  async deleteResult(id: string): Promise<void> {
    const response = await fetch(`/api/v1/backtest/results/${id}`, {
      method: 'DELETE'
    })
    if (!response.ok) throw new Error('Failed to delete backtest result')
  }

  async compareResults(resultIds: string[]): Promise<BacktestComparison> {
    const response = await fetch('/api/v1/backtest/compare', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ resultIds })
    })
    if (!response.ok) throw new Error('Failed to compare backtest results')
    return response.json()
  }

  async exportResults(resultIds: string[]): Promise<void> {
    const response = await fetch('/api/v1/backtest/export', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ resultIds })
    })
    if (!response.ok) throw new Error('Failed to export backtest results')
    
    // Handle file download
    const blob = await response.blob()
    const url = window.URL.createObjectURL(blob)
    const a = document.createElement('a')
    a.href = url
    a.download = `backtest-results-${new Date().toISOString()}.csv`
    document.body.appendChild(a)
    a.click()
    window.URL.revokeObjectURL(url)
    document.body.removeChild(a)
  }

  async getStrategyTemplates(): Promise<StrategyTemplate[]> {
    const response = await fetch('/api/v1/backtest/strategies')
    if (!response.ok) throw new Error('Failed to fetch strategy templates')
    return response.json()
  }

  async validateStrategy(code: string): Promise<{
    valid: boolean
    errors?: string[]
    warnings?: string[]
  }> {
    const response = await fetch('/api/v1/backtest/validate-strategy', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ code })
    })
    if (!response.ok) throw new Error('Failed to validate strategy')
    return response.json()
  }

  async optimizeParameters(
    configId: string,
    parameters: Record<string, { min: number; max: number; step: number }>
  ): Promise<{
    optimizationId: string
  }> {
    const response = await fetch(`/api/v1/backtest/optimize/${configId}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ parameters })
    })
    if (!response.ok) throw new Error('Failed to start parameter optimization')
    return response.json()
  }

  async getOptimizationResults(optimizationId: string): Promise<{
    status: BacktestStatus
    progress: number
    results?: Array<{
      parameters: Record<string, number>
      sharpeRatio: number
      totalReturn: number
      maxDrawdown: number
    }>
  }> {
    const response = await fetch(`/api/v1/backtest/optimize/${optimizationId}/results`)
    if (!response.ok) throw new Error('Failed to fetch optimization results')
    return response.json()
  }

  // WebSocket connection for real-time backtest progress
  connectBacktestProgress(
    backtestId: string,
    onProgress: (progress: BacktestProgress) => void,
    onError?: (error: Error) => void
  ): () => void {
    const ws = new WebSocket(
      `${window.location.protocol === 'https:' ? 'wss:' : 'ws:'}//${window.location.host}/ws/backtest/${backtestId}`
    )
    
    ws.onmessage = (event) => {
      try {
        const progress = JSON.parse(event.data)
        onProgress(progress)
      } catch (error) {
        onError?.(new Error('Failed to parse backtest progress'))
      }
    }
    
    ws.onerror = () => {
      onError?.(new Error('WebSocket connection error'))
    }
    
    return () => ws.close()
  }
}

export const backtestService = new BacktestService()
